datafusion/physical-plan/src/lib.rs

```diff
@@ -121,21 +121,23 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     fn as_any(&self) -> &dyn Any;
 
     /// Get the schema for this execution plan
-    fn schema(&self) -> SchemaRef;
+    fn schema(&self) -> SchemaRef {
+        self.cache().schema().clone()
+    }
+
+    fn cache(&self) -> &PlanPropertiesCache;
 
     /// Specifies how the output of this `ExecutionPlan` is split into
     /// partitions.
-    fn output_partitioning(&self) -> Partitioning;
+    fn output_partitioning(&self) -> &Partitioning {
```

Comment:
Given that `fn cache()` is required, it seems to me there is a potential in the API for conflicting implementations (e.g. a user could return a default `PlanProperties` but forget to remove `output_partitioning`

This I suggest removing the functions like `output_partitioning` from the `ExecutionPlan` trait.

You could probably avoid having to make a bunch of changes by moving the functions into a new trait like

```rust
pub trait ExecutionPlanExt {
    fn execution_mode(&self) -> ExecutionMode;
...
}

// implement so current call sites of plan.execution_mode() still work:
impl ExecutionPlanExt for &dyn ExecutionPlan {
    fn execution_mode(&self) -> ExecutionMode;
       self.cache().exec_mode
    }
 
```

Reply:
I think @mustafasrepo addressed all the reviews including naming, this is the only one left. I don't think I fully understand your point, can you explain a little more?

Specifically, the only difference I see with @mustafasrepo's design (i.e. default implementations) and using a new extension trait is that the former leaves the door open for a `ExecutionPlan` implementation to override the simple "get the property from cache" behavior for any of the functions like `output_partitioning`. I think this was by design, in case a downstream requirement forces implementors to add some custom behavior there.

Do you think that may be undesirable (or have some possibly harmful results)?

Reply:
> Specifically, the only difference I see with @mustafasrepo's design (i.e. default implementations) and using a new extension trait is that the former leaves the door open for a `ExecutionPlan` implementation to override the simple "get the property from cache" behavior for any of the functions like `output_partitioning`. I think this was by design, in case a downstream requirement forces implementors to add some custom behavior there.
> 
> Do you think that may be undesirable (or have some possibly harmful results)?

The problem I see is that it would be possible for `plan.properties().execution_mode` to be inconsistent with
 `plan.execution_mode()`. 

What I was imagining was the upgrade scenario of existing implementation of `ExecutionPlan` such as we have in `InfluxDB` 

1. The person doing the upgrade sees "I have to implement a new function named `properties()`"
2. Simply implements it by returning a default `PlanProperties()` function but leaves the `execution_mode` impl unchanged

Now I can imagine some code in the future assumes that since all `ExecutionPlan` impls in DataFusion correctly implement `properties()` that it is safe to assume that `plan.properties().execution_mode` matches `plan.execution_mode()`.

Maybe this would have been less likely when the name was cache() 

I don't feel super strongly about this issue but I do think it is a "foot gun" in the API

Reply:
I see. This makes sense -- @mustafasrepo, let's create an extension trait like @alamb described

Reply:
I have applied your suggesion @alamb. Thanks for the suggestion. It is very neat. I thought about this possibility, However couldn't come up with a way to prevent it.